lib/config/applyrule.hpp

```diff
@@ -15,11 +16,42 @@ namespace icinga
 /**
  * @ingroup config
  */
-class ApplyRule
+class ApplyRule : public SharedObject
 {
 public:
+	DECLARE_PTR_TYPEDEFS(ApplyRule);
+
+	struct PerHost
+	{
+		std::vector<ApplyRule::Ptr> ForHost;
+		std::unordered_map<String /* service */, std::vector<ApplyRule::Ptr>> ForServices;
+	};
+
+	struct PerTypes
+	{
+		std::vector<ApplyRule::Ptr> Regular;
+		std::unordered_map<String /* host */, PerHost> Targeted;
+	};
+
+	/*
+	 * m_Rules[T::TypeInstance.get()][C::TypeInstance.get()].Regular
+	 * contains all apply rules like apply T "x" to C { ... }
+	 * which don't target only specific checkables, i.e. not just
+	 * assign where host.name == "H" || host.name == "h".
```

Comment:
I find the "i.e." part strange as this only covers hosts. I'd just put this one last and make it say something like "contains all other apply rules, these will be evaluated on all C objects" (why C though? child? then it would be the wrong way around)

Reply:
C for Checkable